src/Features/Core/Portable/QuickInfo/QuickInfoContext.cs

```diff
@@ -4,40 +4,87 @@
 
 using System;
 using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 
 namespace Microsoft.CodeAnalysis.QuickInfo
 {
     /// <summary>
     /// The context presented to a <see cref="QuickInfoProvider"/> when providing quick info.
     /// </summary>
-    internal sealed class QuickInfoContext
+    internal abstract class QuickInfoContext
     {
         /// <summary>
-        /// The document that quick info was requested within.
+        /// The semantic model that quick info was requested within.
         /// </summary>
-        public Document Document { get; }
+        public SemanticModel SemanticModel { get; }
```

Comment:
hrmm... i don't love this change.  I think the intent of the previous way was so that we could eventually make this system public.  In that sense the core QuickInfoContext would be language neutral, and languages would implement the provider.  However, langauges could also return subtypes that added this information to pass along.

Reply:
i think what we probably need (and pulling in @jasonmalinowski on how we should structure this) is some sort of lower level API, while keeping the core IQuckInfo api Document-Centric.  Maybe ICompilerQuickInfo/ICompilationQuickInfo (or some other name that isn't quick info), which is C#/VB speific and can thus refer to these lower-level, C#/VB-specific, types.  

We'd then layer roslyn's real IQuickInfo type on top of this lower level type that contained our language logic.  This owuld allow us to have a service we could use for lsif/and C#/VB-ide, while still keeping around the bones of an editor-level API we would want for our other supported languages. 

Happy to talk offline aboe this.

Reply:
@CyrusNajmabadi I think all the changes you mentioned for future public API exposure can be done as a follow-up change where we define the required public document and compilation centric interfaces and have the QuickInfoContext sub-types implement the appropriate interfaces. Each provider would need to somehow declare if it works on document model OR compilation model and then the driver can invoke the appropriate overload with the correct interface as the parameter. I don't think the current change will prevent such a future refactoring.

Reply:
But won't this change literally break typescript quick info?  They certainly can't use our semantic entities.  They need the Lang agnostic api to provide quickinfo through. 

All I'm saying is that our C#/vb impl of the Lang neutral api call into a semantic api that does use these roslyn Lang specific semantic entities as currency.

Reply:
Ah, I did not know TS already used this - nothing in external access layer referenced it. I get your point though, let me tweak this tomorrow.

Reply:
I have combined everything into QuickInfoContext for now. In future, when we want the API to be public, we can split out the type or introduce interfaces as appropriate.